test/jdk/javax/swing/plaf/basic/BasicDirectoryModel/ConcurrentModification.java

```diff
@@ -0,0 +1,237 @@
+    /** Initial number of files. */
+    private static final long NUMBER_OF_FILES = 50;
+    /** Maximum number of files created on a timer tick. */
+    private static final long LIMIT_FILES = 10;
+
+    /** Timer period (delay) for creating new files. */
+    private static final long TIMER_PERIOD = 250;
+
+    /**
+     * Number of threads running {@code fileChooser.rescanCurrentDirectory()}.
+     */
+    private static final int NUMBER_OF_THREADS = 5;
+    /** Number of repeated calls to {@code rescanCurrentDirectory}. */
+    private static final int NUMBER_OF_REPEATS = 2_000;
+    /** Maximum amount a thread waits before initiating rescan. */
+    private static final long LIMIT_SLEEP = 100;
+
+
+    /** The barrier to start all the scanner threads simultaneously. */
+    private static final CyclicBarrier start = new CyclicBarrier(NUMBER_OF_THREADS);
+    /** The barrier to wait for all the scanner threads to complete, plus main thread. */
+    private static final CyclicBarrier end = new CyclicBarrier(NUMBER_OF_THREADS + 1);
+
+    /** List of scanner threads. */
+    private static final List<Thread> threads = new ArrayList<>(NUMBER_OF_THREADS);
+
+    /**
+     * Stores an exception caught by any of the threads.
+     * If more exceptions are caught, they're added as suppressed exceptions.
+     */
+    private static final AtomicReference<Throwable> exception =
+            new AtomicReference<>();
+
+
+    public static void main(String[] args) throws Throwable {
+        try {
+            // Start the test in its own thread group to catch and handle
+            // all thrown exceptions, in particular in
+            // BasicDirectoryModel.FilesLoader which is created by Swing.
+            ThreadGroup threadGroup = new ConcurrentModification();
+            Thread runner = new Thread(threadGroup,
+                                       ConcurrentModification::wrapper,
+                                       "Test Runner");
+            runner.start();
+            runner.join();
+        } catch (Throwable throwable) {
+            handleException(throwable);
+        }
+
+        if (exception.get() != null) {
+            throw exception.get();
+        }
+    }
+
+    private static void wrapper() {
+        final long timeStart = System.currentTimeMillis();
+        try {
+            runTest(timeStart);
+        } catch (Throwable throwable) {
+            handleException(throwable);
+        } finally {
+            System.out.printf("Duration: %,d\n",
+                              (System.currentTimeMillis() - timeStart));
+        }
+    }
+
+    private static void runTest(final long timeStart) throws Throwable {
+        final Path temp = Files.createDirectory(Paths.get("fileChooser-concurrency-" + timeStart));
+
+        final Timer timer = new Timer("File creator");
+
+        try {
+            createFiles(temp);
+
+            final JFileChooser fc = new JFileChooser(temp.toFile());
```

Comment:
it should be created on EDT, just in case.

Reply:
I don't think it matters here… The test does not rely on even processing on EDT, no state of `JFileChooser` is modified except for calling `rescanCurrentDirectory`. Technically, I am not allowed to do so either. But if I call `rescanCurrentDirectory` from EDT only, the test becomes void.

Reply:
I think it will fire the property change events via firePropertyChange from the constructor and it is better to do that on EDT.

Reply:
I guess it does fire property change events on the current thread rather than EDT in this case. I still don't think it matters in this case. After the `JFileChooser` is created, it is accessed on the Scanner threads only, and only to call its `rescanCurrentDirectory` method. This call, in its turn, creates a new background thread which accesses the fields and methods of `JFileChooser` as well as `FileSystemView`.

Instantiating `JFileChooser` on EDT adds complexity for no benefit. The `JFileChooser` is still accessed concurrently from several threads, including Swing internal thread.

Reply:
>Instantiating JFileChooser on EDT adds complexity for no benefit. The JFileChooser is still accessed concurrently from several threads, including Swing internal thread.

But still, this is a common requirement to create UI components on EDT, and that will replicate more or less how it will work in the application.

Reply:
> > Instantiating JFileChooser on EDT adds complexity for no benefit. The JFileChooser is still accessed concurrently from several threads, including Swing internal thread.
> 
> But still, this is a common requirement to create UI components on EDT, and that will replicate more or less how it will work in the application.

And it is a common requirement ***not** to call `fileChooser.rescanCurrentDirectory` concurrently* in 5 different threads. So it does not apply.

Reply:
In general, I agree all Swing components are to be created and accessed on EDT.

However, there are cases where instantiating and testing is safe on main thread. It is when the component is not shown on the screen, its state does not depend on any events dispatched by the event queue. There are such tests in the JDK code base.

I think this test is one of such cases where `JFileChooser` can be (safely) created on the current thread. The test does not display the file chooser, its state does not depend on receiving input events and handling them.

Essentially, the test verifies that *the implementation of `JFileChooser.rescanCurrentDirectory` is thread-safe*. It wasn't. Two bugs were resolved to make the implementation thread-safe: [JDK-8323670](https://bugs.openjdk.org/browse/JDK-8323670) and [JDK-8325179](https://bugs.openjdk.org/browse/JDK-8325179).

Reply:
> However, there are cases where instantiating and testing is safe on main thread.

That is my point, make it less safe - when the component is created on on thread:EDT and then for some reason accessed on a different thread.

Reply:
> > However, there are cases where instantiating and testing is safe on main thread.
> 
> That is my point, make it less safe - when the component is created on one thread:EDT and then for some reason accessed on a different thread, the rescanCurrentDirectory should still work.

I got your idea!

However, I still see no benefit for doing it.

Yes, when `JFileChooser` is instantiated, it creates its UI delegates and the default model, `BasicDirectoryModel`. The constructor of the model calls `validateFileCache` which starts the background thread and will update `fileCache`.

It does not make the initialisation less thread-safe: the same two threads are involved — `FilesLoader` and EDT. Yet there's no race yet…

The `ConcurrentModificationException` is thrown when the background `FilesLoader` is iterating over `fileCache` while EDT runs `DoChangeContents.run` which adds elements to `fileCache` or removes elements from it.

By the time the `Scanner` threads start, a race becomes possible. If the initial `FilesLoader` completes and posts the `DoChangeContents` object to EDT, the `FilesLoader` threads created by the scanners will race… but `fileCache` is empty initially, thus iterating over it is very quick.

Therefore getting `ConcurrentModificationException` is unlikely until `fileCache` contains a list of files. This state is reached later in the timeline of the test.

The initial `FilesLoader` thread that's started when `JFileChooser` is instantiated plays a role too… Yet it's running concurrently with the `Scanner` threads whether the `JFileChooser` object is created on the current thread or on EDT.

---

Taking the above into account, instantiating `JFileChooser` adds complexity to the test but brings no benefits.

Reply:
ok, I agree with you.